File: server/src/test/java/de/zalando/zally/rules/RulesValidatorTest.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Swagger;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class RulesValidatorTest {

    private static final Violation DUMMY_VIOLATION_1 = new Violation("dummy1", "dummy", ViolationType.MUST, "dummy");
    private static final Violation DUMMY_VIOLATION_2 = new Violation("dummy2", "dummy", ViolationType.MUST, "dummy");
    private static final Violation DUMMY_VIOLATION_3 = new Violation("dummy3", "dummy", ViolationType.MUST, "dummy");
    private static final Violation DUMMY_VIOLATION_PATH_A = new Violation("dummy3", "dummy", ViolationType.MUST, "dummy", "a");
    private static final Violation DUMMY_VIOLATION_PATH_Z = new Violation("dummy3", "dummy", ViolationType.MUST, "dummy", "z");

    private static class OneViolationRule implements Rule {

        @Override
        public List<Violation> validate(Swagger swagger) {
            return Arrays.asList(DUMMY_VIOLATION_1);
        }
    }

    private static class TwoViolationRule implements Rule {

        @Override
        public List<Violation> validate(Swagger swagger) {
            return Arrays.asList(DUMMY_VIOLATION_2, DUMMY_VIOLATION_3);
        }
    }

    private static class PathViolationRule implements Rule {

        @Override
        public List<Violation> validate(Swagger swagger) {
            return Arrays.asList(DUMMY_VIOLATION_PATH_Z, DUMMY_VIOLATION_PATH_A);
        }
    }

    @Test
    public void shouldReturnEmptyViolationsListWithoutRules() {
        RulesValidator validator = new RulesValidator(Collections.emptyList());
        assertThat(validator.validate(new Swagger())).isEmpty();
    }

    @Test
    public void shouldReturnOneViolation() {
        RulesValidator validator = new RulesValidator(Arrays.asList(new OneViolationRule()));
        assertThat(validator.validate(new Swagger())).containsOnly(DUMMY_VIOLATION_1);
    }

    @Test
    public void shouldCollectViolationsOfAllRules() {
        RulesValidator validator = new RulesValidator(Arrays.asList(new OneViolationRule(), new TwoViolationRule()));
        assertThat(validator.validate(new Swagger())).containsOnly(DUMMY_VIOLATION_1, DUMMY_VIOLATION_2, DUMMY_VIOLATION_3);
    }

    @Test
    public void shouldSortViolationsAlphabeticallyAndWithoutPathOnTop() {
        RulesValidator validator = new RulesValidator(Arrays.asList(new OneViolationRule(), new PathViolationRule()));
        assertThat(validator.validate(new Swagger())).containsExactly(DUMMY_VIOLATION_1, DUMMY_VIOLATION_PATH_A, DUMMY_VIOLATION_PATH_Z);
    }
}

File: server/src/test/java/de/zalando/zally/rules/RulesValidatorTest.kt
package de.zalando.zally.rules

import de.zalando.zally.Violation
import de.zalando.zally.ViolationType
import io.swagger.models.Swagger
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class RulesValidatorTest {

    val DUMMY_VIOLATION_1 = Violation(TestRule(null), "dummy1", "dummy", ViolationType.SHOULD, "dummy", listOf("x", "y", "z"))
    val DUMMY_VIOLATION_2 = Violation(TestRule(null), "dummy2", "dummy", ViolationType.COULD, "dummy", listOf())
    val DUMMY_VIOLATION_3 = Violation(TestRule(null), "dummy3", "dummy", ViolationType.MUST, "dummy", listOf("a"))

    val swaggerContent = javaClass.classLoader.getResource("fixtures/api_spp.json").readText(Charsets.UTF_8)

    class TestRule(val result: Violation?) : AbstractRule() {
        override fun validate(swagger: Swagger): Violation? = result
    }

    @Test
    fun shouldReturnEmptyViolationsListWithoutRules() {
        val validator = RulesValidator(emptyList())
        assertThat(validator.validate(swaggerContent)).isEmpty()
    }

    @Test
    fun shouldReturnOneViolation() {
        val violations = listOf(DUMMY_VIOLATION_1)
        val validator = RulesValidator(violations.map { TestRule(it) })
        assertThat(validator.validate(swaggerContent)).hasSameElementsAs(violations)
    }

    @Test
    fun shouldCollectViolationsOfAllRules() {
        val violations = listOf(DUMMY_VIOLATION_1, DUMMY_VIOLATION_2)
        val validator = RulesValidator(violations.map { TestRule(it) })
        assertThat(validator.validate(swaggerContent)).hasSameElementsAs(violations)
    }

    @Test
    fun shouldSortViolationsByViolationType() {
        val violations = listOf(DUMMY_VIOLATION_1, DUMMY_VIOLATION_2, DUMMY_VIOLATION_3)
        val validator = RulesValidator(violations.map { TestRule(it) })
        assertThat(validator.validate(swaggerContent))
                .containsExactly(DUMMY_VIOLATION_3, DUMMY_VIOLATION_1, DUMMY_VIOLATION_2)
    }
}
